rfcs/0013-changes-lifecyle.md

```diff
@@ -0,0 +1,348 @@
+
+If a feature requires or allows new configurations, these configurations should be optional, and as soon as confirmed the feature is enabled, it should verify if the required configuration is available.
+
+```
+> forc ... --experimental a_feature --a-feature-option 1
+```
+
+## Lexer and Parser
+
+To allow as user-friendly error messages as possible, in all possible cases, we want both the lexer and the parser to parse the new syntax even with the feature off.
+
+After that, the lexer and the parser will mark that an experimental feature was lexed/parsed; and a check will guarantee that no disabled experimental was parsed.
+
+The error message will have a message explaining that the feature is experimental and a GitHub link for more details on the stabilization lifecycle.
+
+```rust
+// always parse new syntax
+let new_syntax = parse_new_syntax();
+if ctx.experimental.is_disabled(Features::NEW_FEATURE) {
+   handler.emit_error(...);
+}
+```
+
+## Formatting
+
+Formatting should always format new syntax, even when the flag is off. To avoid breaking other parts of the code, or even worse, removing code.
+
+## LSP
+
+LSP can take advantage of specific error messages, and suggest users to enable the corresponding feature.
+
+## CST, AST, and Typed Tree
+
+All trees must always support new features, which means that new nodes will always exist. Their specific behavior, desugaring, etc., will be gated by the experimental feature.
+
+More specifically, this means that variants should not be behind compiler flags.
+
+## Tests
+
+To allow these experimental features to be tested, it is necessary to specify which configurations each test must run.
+Today, by default, a test runs with the "default compiler configuration", but it is necessary to allow tests to
+run with any combination of compiler configuration.
+
+To allow complete tests of any compiler configuration, any configuration that today resides inside `test.toml`
+needs to be configured for each configuration.
+
+An example is the `encoding v1` which has different inputs and outputs from `encoding v0` when testing a script
+that takes arguments.
+
+The easiest way to approach this seems to be to support multiple `test.toml` with a suffix to differentiate them.
+
+```
+test.toml
+test.feature_a.toml
+```
+
+To avoid duplications `test.feature_a.toml` can inherit properties from `test.toml`.
+And some new properties can be created to allow the configuration of the compiler:
+
+```
+[environment_variables]
+SOME_VAR = "1"
+
+[forc]
+cli = "--experimental feature_a,feature_b"
+```
+
+The same strategy can be used for snapshot tests. We can use multiple `snapshot.toml` and create different
+snapshots using the file suffix. So `snapshot.feature_a.toml` will generate `snapshot@feature_a.snap`.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+This will increase the complexity of the compiler. Not all flags used to compile end up in the JSON ABI, or other outputs. Which can make reproducibility harder.
```

Comment:
this may impact abi verification tools (cc @sdankel)

Reply:
Surely such tools should only use stable compiler features. We may want to extend support which requires recoding what flags are used on top of the compiler version, but we should not encourage people to use experimental features. In fact this should be more discouraged than it is in Rust given our environment.

Reply:
I vote for this to be dealt in another RFC.